File: backend/services/bank_reconciliation_service.go
package services

import (
	"app-sistem-akuntansi/models"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"gorm.io/gorm"
)

type BankReconciliationService struct {
	db *gorm.DB
}

func NewBankReconciliationService(db *gorm.DB) *BankReconciliationService {
	return &BankReconciliationService{db: db}
}

// ========== SNAPSHOT OPERATIONS ==========

// CreateSnapshotRequest - Request untuk membuat snapshot
type CreateSnapshotRequest struct {
	CashBankID uint   `json:"cash_bank_id" binding:"required"`
	Period     string `json:"period" binding:"required"` // Format: YYYY-MM
	Notes      string `json:"notes"`
}

// GenerateSnapshot - Membuat snapshot transaksi untuk periode tertentu
func (s *BankReconciliationService) GenerateSnapshot(req CreateSnapshotRequest, userID uint) (*models.BankReconciliationSnapshot, error) {
	// Validate cash bank exists
	var cashBank models.CashBank
	if err := s.db.First(&cashBank, req.CashBankID).Error; err != nil {
		return nil, errors.New("cash bank account not found")
	}

	// Parse period
	periodTime, err := time.Parse("2006-01", req.Period)
	if err != nil {
		return nil, errors.New("invalid period format, use YYYY-MM")
	}

	// Get transactions for the period
	startDate := periodTime
	endDate := periodTime.AddDate(0, 1, 0).Add(-time.Second)

	var transactions []models.CashBankTransaction
	if err := s.db.Where("cash_bank_id = ? AND transaction_date >= ? AND transaction_date <= ?",
		req.CashBankID, startDate, endDate).
		Order("transaction_date ASC, id ASC").
		Find(&transactions).Error; err != nil {
		return nil, err
	}

	// Calculate balances
	var openingBalance, closingBalance, totalDebit, totalCredit float64
	
	// Get opening balance (balance before period start)
	if err := s.db.Raw(`
		SELECT COALESCE(balance_after, 0) as balance
		FROM cash_bank_transactions
		WHERE cash_bank_id = ? AND transaction_date < ?
		ORDER BY transaction_date DESC, id DESC
		LIMIT 1
	`, req.CashBankID, startDate).Scan(&openingBalance).Error; err != nil {
		// If no previous transactions, opening balance is 0
		openingBalance = 0
	}

	// Calculate totals
	for _, tx := range transactions {
		if tx.Amount > 0 {
			totalDebit += tx.Amount
		} else {
			totalCredit += -tx.Amount
		}
	}

	// Closing balance is opening + debit - credit
	closingBalance = openingBalance + totalDebit - totalCredit

	// Create snapshot
	snapshot := models.BankReconciliationSnapshot{
		CashBankID:       req.CashBankID,
		Period:           req.Period,
		SnapshotDate:     time.Now(),
		GeneratedBy:      userID,
		OpeningBalance:   openingBalance,
		ClosingBalance:   closingBalance,
		TotalDebit:       totalDebit,
		TotalCredit:      totalCredit,
		TransactionCount: len(transactions),
		Notes:            req.Notes,
		Status:           "ACTIVE",
	}

	// Begin transaction
	tx := s.db.Begin()

	// Save snapshot
	if err := tx.Create(&snapshot).Error; err != nil {
		tx.Rollback()
		return nil, err
	}

	// Save transaction snapshots
	for _, transaction := range transactions {
		snapshotTx := models.ReconciliationTransactionSnapshot{
			SnapshotID:      snapshot.ID,
			TransactionID:   transaction.ID,
			TransactionDate: transaction.TransactionDate,
			ReferenceType:   transaction.ReferenceType,
			ReferenceID:     transaction.ReferenceID,
			Amount:          transaction.Amount,
			BalanceAfter:    transaction.BalanceAfter,
			Notes:           transaction.Notes,
			CreatedBy:       userID,
		}

		// Determine debit/credit
		if transaction.Amount > 0 {
			snapshotTx.DebitAmount = transaction.Amount
			snapshotTx.CreditAmount = 0
		} else {
			snapshotTx.DebitAmount = 0
			snapshotTx.CreditAmount = -transaction.Amount
		}

		if err := tx.Create(&snapshotTx).Error; err != nil {
			tx.Rollback()
			return nil, err
		}
	}

	// Generate hash for integrity
	hash, err := s.generateSnapshotHash(&snapshot, transactions)
	if err != nil {
		tx.Rollback()
		return nil, err
	}
	snapshot.DataHash = hash

	if err := tx.Model(&snapshot).Update("data_hash", hash).Error; err != nil {
		tx.Rollback()
		return nil, err
	}

	// Commit transaction
	if err := tx.Commit().Error; err != nil {
		return nil, err
	}

	// Load relations
	s.db.Preload("CashBank").Preload("GeneratedByUser").Preload("Transactions").First(&snapshot, snapshot.ID)

	return &snapshot, nil
}

// generateSnapshotHash - Generate SHA-256 hash untuk snapshot integrity
func (s *BankReconciliationService) generateSnapshotHash(snapshot *models.BankReconciliationSnapshot, transactions []models.CashBankTransaction) (string, error) {
	// Create a composite structure for hashing
	hashData := struct {
		CashBankID       uint
		Period           string
		OpeningBalance   float64
		ClosingBalance   float64
		TotalDebit       float64
		TotalCredit      float64
		TransactionCount int
		Transactions     []models.CashBankTransaction
	}{
		CashBankID:       snapshot.CashBankID,
		Period:           snapshot.Period,
		OpeningBalance:   snapshot.OpeningBalance,
		ClosingBalance:   snapshot.ClosingBalance,
		TotalDebit:       snapshot.TotalDebit,
		TotalCredit:      snapshot.TotalCredit,
		TransactionCount: snapshot.TransactionCount,
		Transactions:     transactions,
	}

	// Convert to JSON
	jsonData, err := json.Marshal(hashData)
	if err != nil {
		return "", err
	}

	// Generate SHA-256 hash
	hash := sha256.Sum256(jsonData)
	return hex.EncodeToString(hash[:]), nil
}

// LockSnapshot - Lock snapshot untuk prevent modifications
func (s *BankReconciliationService) LockSnapshot(snapshotID uint, userID uint) error {
	now := time.Now()
	return s.db.Model(&models.BankReconciliationSnapshot{}).
		Where("id = ?", snapshotID).
		Updates(map[string]interface{}{
			"is_locked": true,
			"locked_at": now,
			"locked_by": userID,
		}).Error
}

// GetSnapshots - Get all snapshots for a cash bank account
func (s *BankReconciliationService) GetSnapshots(cashBankID uint) ([]models.BankReconciliationSnapshot, error) {
	var snapshots []models.BankReconciliationSnapshot
	err := s.db.Where("cash_bank_id = ?", cashBankID).
		Preload("CashBank").
		Preload("GeneratedByUser").
		Order("snapshot_date DESC").
		Find(&snapshots).Error
	return snapshots, err
}

// GetSnapshotByID - Get snapshot by ID with all details
func (s *BankReconciliationService) GetSnapshotByID(id uint) (*models.BankReconciliationSnapshot, error) {
	var snapshot models.BankReconciliationSnapshot
	err := s.db.Preload("CashBank").
		Preload("GeneratedByUser").
		Preload("Transactions").
		First(&snapshot, id).Error
	return &snapshot, err
}

// ========== RECONCILIATION OPERATIONS ==========

// CreateReconciliationRequest - Request untuk membuat rekonsiliasi
type CreateReconciliationRequest struct {
	CashBankID           uint   `json:"cash_bank_id" binding:"required"`
	Period               string `json:"period" binding:"required"`
	BaseSnapshotID       uint   `json:"base_snapshot_id" binding:"required"`
	ComparisonSnapshotID *uint  `json:"comparison_snapshot_id"` // Optional, akan di-generate jika null
	Notes                string `json:"notes"`
}

// PerformReconciliation - Melakukan rekonsiliasi antara base snapshot dan current data
func (s *BankReconciliationService) PerformReconciliation(req CreateReconciliationRequest, userID uint) (*models.BankReconciliation, error) {
	// Get base snapshot
	baseSnapshot, err := s.GetSnapshotByID(req.BaseSnapshotID)
	if err != nil {
		return nil, errors.New("base snapshot not found")
	}

	// If comparison snapshot ID not provided, generate new snapshot
	var comparisonSnapshot *models.BankReconciliationSnapshot
	if req.ComparisonSnapshotID == nil {
		// Generate new snapshot for current state
		newSnapshot, err := s.GenerateSnapshot(CreateSnapshotRequest{
			CashBankID: req.CashBankID,
			Period:     req.Period,
			Notes:      "Auto-generated for reconciliation",
		}, userID)
		if err != nil {
			return nil, err
		}
		comparisonSnapshot = newSnapshot
		req.ComparisonSnapshotID = &newSnapshot.ID
	} else {
		comparisonSnapshot, err = s.GetSnapshotByID(*req.ComparisonSnapshotID)
		if err != nil {
			return nil, errors.New("comparison snapshot not found")
		}
	}

	// Compare snapshots
	differences, err := s.compareSnapshots(baseSnapshot, comparisonSnapshot)
	if err != nil {
		return nil, err
	}

	// Calculate variance
	variance := comparisonSnapshot.ClosingBalance - baseSnapshot.ClosingBalance
	transactionVariance := comparisonSnapshot.TransactionCount - baseSnapshot.TransactionCount
	isBalanced := variance == 0 && transactionVariance == 0 && len(differences) == 0

	// Create reconciliation record
	reconciliation := models.BankReconciliation{
		CashBankID:              req.CashBankID,
		Period:                  req.Period,
		BaseSnapshotID:          req.BaseSnapshotID,
		ComparisonSnapshotID:    req.ComparisonSnapshotID,
		ReconciliationDate:      time.Now(),
		ReconciliationBy:        userID,
		BaseBalance:             baseSnapshot.ClosingBalance,
		CurrentBalance:          comparisonSnapshot.ClosingBalance,
		Variance:                variance,
		BaseTransactionCount:    baseSnapshot.TransactionCount,
		CurrentTransactionCount: comparisonSnapshot.TransactionCount,
		TransactionVariance:     transactionVariance,
		MissingTransactions:     s.countDifferencesByType(differences, "MISSING"),
		AddedTransactions:       s.countDifferencesByType(differences, "ADDED"),
		ModifiedTransactions:    s.countDifferencesByType(differences, "MODIFIED"),
		Status:                  "PENDING",
		IsBalanced:              isBalanced,
		Notes:                   req.Notes,
	}

	// Begin transaction
	tx := s.db.Begin()

	// Save reconciliation
	if err := tx.Create(&reconciliation).Error; err != nil {
		tx.Rollback()
		return nil, err
	}

	// Save differences
	for _, diff := range differences {
		diff.ReconciliationID = reconciliation.ID
		if err := tx.Create(&diff).Error; err != nil {
			tx.Rollback()
			return nil, err
		}
	}

	// Commit transaction
	if err := tx.Commit().Error; err != nil {
		return nil, err
	}

	// Load relations
	s.db.Preload("CashBank").
		Preload("BaseSnapshot").
		Preload("ComparisonSnapshot").
		Preload("ReconciliationByUser").
		Preload("Differences").
		First(&reconciliation, reconciliation.ID)

	return &reconciliation, nil
}

// compareSnapshots - Bandingkan 2 snapshot dan temukan perbedaan
func (s *BankReconciliationService) compareSnapshots(base, comparison *models.BankReconciliationSnapshot) ([]models.ReconciliationDifference, error) {
	var differences []models.ReconciliationDifference

	// Create maps for quick lookup
	baseMap := make(map[uint]models.ReconciliationTransactionSnapshot)
	comparisonMap := make(map[uint]models.ReconciliationTransactionSnapshot)

	for _, tx := range base.Transactions {
		baseMap[tx.TransactionID] = tx
	}

	for _, tx := range comparison.Transactions {
		comparisonMap[tx.TransactionID] = tx
	}

	// Find missing transactions (in base but not in comparison)
	for txID, baseTx := range baseMap {
		if _, exists := comparisonMap[txID]; !exists {
			differences = append(differences, models.ReconciliationDifference{
				DifferenceType:   "MISSING",
				Severity:         "HIGH",
				BaseTransactionID: &baseTx.TransactionID,
				Description:      fmt.Sprintf("Transaction %d missing in current data", txID),
				Status:           "PENDING",
			})
		}
	}

	// Find added transactions (in comparison but not in base)
	for txID, compTx := range comparisonMap {
		if _, exists := baseMap[txID]; !exists {
			differences = append(differences, models.ReconciliationDifference{
				DifferenceType:       "ADDED",
				Severity:             "MEDIUM",
				CurrentTransactionID: &compTx.TransactionID,
				Description:          fmt.Sprintf("Transaction %d added after snapshot", txID),
				Status:               "PENDING",
			})
		}
	}

	// Find modified transactions
	for txID, baseTx := range baseMap {
		if compTx, exists := comparisonMap[txID]; exists {
			// Check if amount changed
			if baseTx.Amount != compTx.Amount {
				baseTxID := baseTx.TransactionID
				compTxID := compTx.TransactionID
				differences = append(differences, models.ReconciliationDifference{
					DifferenceType:       "AMOUNT_CHANGE",
					Severity:             "CRITICAL",
					BaseTransactionID:    &baseTxID,
					CurrentTransactionID: &compTxID,
					Field:                "amount",
					OldValue:             fmt.Sprintf("%.2f", baseTx.Amount),
					NewValue:             fmt.Sprintf("%.2f", compTx.Amount),
					AmountDifference:     compTx.Amount - baseTx.Amount,
					Description:          fmt.Sprintf("Amount changed for transaction %d", txID),
					Status:               "PENDING",
				})
			}

			// Check if date changed
			if !baseTx.TransactionDate.Equal(compTx.TransactionDate) {
				baseTxID := baseTx.TransactionID
				compTxID := compTx.TransactionID
				differences = append(differences, models.ReconciliationDifference{
					DifferenceType:       "DATE_CHANGE",
					Severity:             "HIGH",
					BaseTransactionID:    &baseTxID,
					CurrentTransactionID: &compTxID,
					Field:                "transaction_date",
					OldValue:             baseTx.TransactionDate.Format("2006-01-02 15:04:05"),
					NewValue:             compTx.TransactionDate.Format("2006-01-02 15:04:05"),
					Description:          fmt.Sprintf("Transaction date changed for transaction %d", txID),
					Status:               "PENDING",
				})
			}
		}
	}

	return differences, nil
}

// countDifferencesByType - Hitung jumlah perbedaan berdasarkan type
func (s *BankReconciliationService) countDifferencesByType(differences []models.ReconciliationDifference, diffType string) int {
	count := 0
	for _, diff := range differences {
		if diff.DifferenceType == diffType {
			count++
		}
	}
	return count
}

// GetReconciliations - Get all reconciliations for a cash bank account
func (s *BankReconciliationService) GetReconciliations(cashBankID uint) ([]models.BankReconciliation, error) {
	var reconciliations []models.BankReconciliation
	err := s.db.Where("cash_bank_id = ?", cashBankID).
		Preload("CashBank").
		Preload("BaseSnapshot").
		Preload("ComparisonSnapshot").
		Preload("ReconciliationByUser").
		Preload("ReviewedByUser").
		Order("reconciliation_date DESC").
		Find(&reconciliations).Error
	return reconciliations, err
}

// GetReconciliationByID - Get reconciliation by ID with all details
func (s *BankReconciliationService) GetReconciliationByID(id uint) (*models.BankReconciliation, error) {
	var reconciliation models.BankReconciliation
	err := s.db.Preload("CashBank").
		Preload("BaseSnapshot").
		Preload("BaseSnapshot.Transactions").
		Preload("ComparisonSnapshot").
		Preload("ComparisonSnapshot.Transactions").
		Preload("ReconciliationByUser").
		Preload("ReviewedByUser").
		Preload("Differences").
		First(&reconciliation, id).Error
	return &reconciliation, err
}

// ApproveReconciliation - Approve reconciliation
func (s *BankReconciliationService) ApproveReconciliation(id uint, userID uint, notes string) error {
	now := time.Now()
	return s.db.Model(&models.BankReconciliation{}).
		Where("id = ?", id).
		Updates(map[string]interface{}{
			"status":        "APPROVED",
			"reviewed_by":   userID,
			"reviewed_at":   now,
			"review_notes":  notes,
			"balance_confirmed": true,
		}).Error
}

// RejectReconciliation - Reject reconciliation
func (s *BankReconciliationService) RejectReconciliation(id uint, userID uint, notes string) error {
	now := time.Now()
	return s.db.Model(&models.BankReconciliation{}).
		Where("id = ?", id).
		Updates(map[string]interface{}{
			"status":       "REJECTED",
			"reviewed_by":  userID,
			"reviewed_at":  now,
			"review_notes": notes,
		}).Error
}

// ========== AUDIT TRAIL OPERATIONS ==========

// LogAuditTrail - Log audit trail untuk perubahan cash bank
func (s *BankReconciliationService) LogAuditTrail(log models.CashBankAuditTrail) error {
	return s.db.Create(&log).Error
}

// GetAuditTrail - Get audit trail for cash bank account
func (s *BankReconciliationService) GetAuditTrail(cashBankID uint, limit int) ([]models.CashBankAuditTrail, error) {
	var logs []models.CashBankAuditTrail
	query := s.db.Where("cash_bank_id = ?", cashBankID).
		Preload("User").
		Preload("ApprovedByUser").
		Order("created_at DESC")
	
	if limit > 0 {
		query = query.Limit(limit)
	}
	
	err := query.Find(&logs).Error
	return logs, err
}

File: frontend/src/services/bankReconciliationService.ts
import api from './api';

// ========== TYPES ==========

export interface BankReconciliationSnapshot {
  id: number;
  cash_bank_id: number;
  period: string;
  snapshot_date: string;
  generated_by: number;
  opening_balance: number;
  closing_balance: number;
  total_debit: number;
  total_credit: number;
  transaction_count: number;
  data_hash: string;
  is_locked: boolean;
  locked_at?: string;
  locked_by?: number;
  notes?: string;
  status: 'ACTIVE' | 'ARCHIVED' | 'SUPERSEDED';
  created_at: string;
  updated_at: string;
  cash_bank?: {
    id: number;
    name: string;
    code: string;
    type: string;
  };
  generated_by_user?: {
    id: number;
    username: string;
  };
  transactions?: ReconciliationTransactionSnapshot[];
}

export interface ReconciliationTransactionSnapshot {
  id: number;
  snapshot_id: number;
  transaction_id: number;
  transaction_date: string;
  reference_type?: string;
  reference_id?: number;
  reference_number?: string;
  amount: number;
  debit_amount: number;
  credit_amount: number;
  balance_after: number;
  description?: string;
  notes?: string;
  created_by: number;
  created_at: string;
}

export interface BankReconciliation {
  id: number;
  reconciliation_number: string;
  cash_bank_id: number;
  period: string;
  base_snapshot_id: number;
  comparison_snapshot_id?: number;
  reconciliation_date: string;
  reconciliation_by: number;
  base_balance: number;
  current_balance: number;
  variance: number;
  base_transaction_count: number;
  current_transaction_count: number;
  transaction_variance: number;
  missing_transactions: number;
  added_transactions: number;
  modified_transactions: number;
  status: 'PENDING' | 'APPROVED' | 'REJECTED' | 'NEEDS_REVIEW';
  reviewed_by?: number;
  reviewed_at?: string;
  review_notes?: string;
  is_balanced: boolean;
  balance_confirmed: boolean;
  notes?: string;
  created_at: string;
  updated_at: string;
  cash_bank?: {
    id: number;
    name: string;
    code: string;
  };
  base_snapshot?: BankReconciliationSnapshot;
  comparison_snapshot?: BankReconciliationSnapshot;
  reconciliation_by_user?: {
    id: number;
    username: string;
  };
  reviewed_by_user?: {
    id: number;
    username: string;
  };
  differences?: ReconciliationDifference[];
}

export interface ReconciliationDifference {
  id: number;
  reconciliation_id: number;
  difference_type: 'MISSING' | 'ADDED' | 'MODIFIED' | 'AMOUNT_CHANGE' | 'DATE_CHANGE';
  severity: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  base_transaction_id?: number;
  current_transaction_id?: number;
  field?: string;
  old_value?: string;
  new_value?: string;
  amount_difference?: number;
  status: 'PENDING' | 'RESOLVED' | 'IGNORED' | 'ESCALATED';
  resolution_notes?: string;
  resolved_by?: number;
  resolved_at?: string;
  description?: string;
  created_at: string;
  updated_at: string;
}

export interface CashBankAuditTrail {
  id: number;
  cash_bank_id: number;
  transaction_id?: number;
  action: 'CREATE' | 'UPDATE' | 'DELETE' | 'VOID' | 'RESTORE';
  entity_type: 'CASH_BANK' | 'TRANSACTION' | 'TRANSFER' | 'DEPOSIT' | 'WITHDRAWAL';
  entity_id: number;
  field_changed?: string;
  old_value?: string;
  new_value?: string;
  reason?: string;
  ip_address?: string;
  user_agent?: string;
  requires_approval: boolean;
  approved_by?: number;
  approved_at?: string;
  approval_status?: 'PENDING' | 'APPROVED' | 'REJECTED';
  user_id: number;
  created_at: string;
  user?: {
    id: number;
    username: string;
  };
  approved_by_user?: {
    id: number;
    username: string;
  };
}

// ========== REQUEST TYPES ==========

export interface CreateSnapshotRequest {
  cash_bank_id: number;
  period: string; // Format: YYYY-MM
  notes?: string;
}

export interface CreateReconciliationRequest {
  cash_bank_id: number;
  period: string;
  base_snapshot_id: number;
  comparison_snapshot_id?: number;
  notes?: string;
}

export interface ApproveRejectRequest {
  notes?: string;
}

// ========== SERVICE CLASS ==========

class BankReconciliationService {
  private baseUrl = '/api/bank-reconciliation';

  // ========== SNAPSHOT OPERATIONS ==========

  async generateSnapshot(request: CreateSnapshotRequest): Promise<BankReconciliationSnapshot> {
    try {
      const response = await api.post(`${this.baseUrl}/snapshots`, request);
      return response.data.data;
    } catch (error) {
      console.error('Error generating snapshot:', error);
      throw error;
    }
  }

  async getSnapshots(cashBankId: number): Promise<BankReconciliationSnapshot[]> {
    try {
      const response = await api.get(`${this.baseUrl}/snapshots`, {
        params: { cash_bank_id: cashBankId, _t: Date.now() },
      });
      return response.data.data;
    } catch (error) {
      console.error('Error fetching snapshots:', error);
      throw error;
    }
  }

  async getSnapshotById(id: number): Promise<BankReconciliationSnapshot> {
    try {
      const response = await api.get(`${this.baseUrl}/snapshots/${id}`, {
        params: { _t: Date.now() },
      });
      return response.data.data;
    } catch (error) {
      console.error('Error fetching snapshot:', error);
      throw error;
    }
  }

  async lockSnapshot(id: number): Promise<void> {
    try {
      await api.post(`${this.baseUrl}/snapshots/${id}/lock`);
    } catch (error) {
      console.error('Error locking snapshot:', error);
      throw error;
    }
  }

  // ========== RECONCILIATION OPERATIONS ==========

  async performReconciliation(request: CreateReconciliationRequest): Promise<BankReconciliation> {
    try {
      const response = await api.post(`${this.baseUrl}/reconcile`, request);
      return response.data.data;
    } catch (error) {
      console.error('Error performing reconciliation:', error);
      throw error;
    }
  }

  async getReconciliations(cashBankId: number): Promise<BankReconciliation[]> {
    try {
      const response = await api.get(`${this.baseUrl}/reconciliations`, {
        params: { cash_bank_id: cashBankId, _t: Date.now() },
      });
      return response.data.data;
    } catch (error) {
      console.error('Error fetching reconciliations:', error);
      throw error;
    }
  }

  async getReconciliationById(id: number): Promise<BankReconciliation> {
    try {
      const response = await api.get(`${this.baseUrl}/reconciliations/${id}`, {
        params: { _t: Date.now() },
      });
      return response.data.data;
    } catch (error) {
      console.error('Error fetching reconciliation:', error);
      throw error;
    }
  }

  async approveReconciliation(id: number, notes?: string): Promise<void> {
    try {
      await api.post(`${this.baseUrl}/reconciliations/${id}/approve`, { notes });
    } catch (error) {
      console.error('Error approving reconciliation:', error);
      throw error;
    }
  }

  async rejectReconciliation(id: number, notes: string): Promise<void> {
    try {
      await api.post(`${this.baseUrl}/reconciliations/${id}/reject`, { notes });
    } catch (error) {
      console.error('Error rejecting reconciliation:', error);
      throw error;
    }
  }

  // ========== AUDIT TRAIL OPERATIONS ==========

  async getAuditTrail(cashBankId: number, limit: number = 100): Promise<CashBankAuditTrail[]> {
    try {
      const response = await api.get(`${this.baseUrl}/audit-trail`, {
        params: { cash_bank_id: cashBankId, limit, _t: Date.now() },
      });
      return response.data.data;
    } catch (error) {
      console.error('Error fetching audit trail:', error);
      throw error;
    }
  }

  async logAuditTrail(log: Partial<CashBankAuditTrail>): Promise<void> {
    try {
      await api.post(`${this.baseUrl}/audit-trail`, log);
    } catch (error) {
      console.error('Error logging audit trail:', error);
      throw error;
    }
  }

  // ========== HELPER FUNCTIONS ==========

  getSeverityColor(severity: string): string {
    switch (severity) {
      case 'CRITICAL':
        return 'red';
      case 'HIGH':
        return 'orange';
      case 'MEDIUM':
        return 'yellow';
      case 'LOW':
        return 'green';
      default:
        return 'gray';
    }
  }

  getStatusColor(status: string): string {
    switch (status) {
      case 'APPROVED':
        return 'green';
      case 'REJECTED':
        return 'red';
      case 'PENDING':
        return 'yellow';
      case 'NEEDS_REVIEW':
        return 'orange';
      default:
        return 'gray';
    }
  }

  getDifferenceTypeLabel(type: string): string {
    switch (type) {
      case 'MISSING':
        return 'Transaction Missing';
      case 'ADDED':
        return 'Transaction Added';
      case 'MODIFIED':
        return 'Transaction Modified';
      case 'AMOUNT_CHANGE':
        return 'Amount Changed';
      case 'DATE_CHANGE':
        return 'Date Changed';
      default:
        return type;
    }
  }

  formatCurrency(amount: number, currency: string = 'IDR'): string {
    return new Intl.NumberFormat('id-ID', {
      style: 'currency',
      currency: currency,
      minimumFractionDigits: 0,
      maximumFractionDigits: 2,
    }).format(amount);
  }

  formatDate(date: string): string {
    return new Date(date).toLocaleDateString('id-ID', {
      year: 'numeric',
      month: 'long',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
    });
  }

  formatPeriod(period: string): string {
    const [year, month] = period.split('-');
    const monthNames = [
      'Januari', 'Februari', 'Maret', 'April', 'Mei', 'Juni',
      'Juli', 'Agustus', 'September', 'Oktober', 'November', 'Desember'
    ];
    return `${monthNames[parseInt(month) - 1]} ${year}`;
  }
}

export default new BankReconciliationService();
